### ISSUE TITLE
Re-adding Mixer Sources

### ISSUE THREAD

**Issue by Sammy1Am** [DSheirer/sdrtrunk#914]
Hi there,

I'd love to have the option to pipe audio into sdrtrunk as a source, so I was digging around to see what it will take to add that feature (... again, since it seems like it used to be there).  I came across [this comment](https://github.com/DSheirer/sdrtrunk/blob/649258bd2bec43d3b8b6de70e361330be954e33c/src/main/java/io/github/dsheirer/source/SourceManager.java#L48-L49) though and figured if I'm going to try to add this functionality I might as well do it as intended by this TODO:

> //TODO: change mixer & recording managers to be models and hand them
        //in via the constructor.  Perform loading outside of this class.

When it says, "models," does it mean like a "tuner model" (i.e. the mixer would be a "tuner" you could choose from), or something that extends `AbstractTableModel` like the `TunerModel` class, or... another meaning of the word "model"?  😛  Looking more closely now I'm guessing it's like the `TunerModel` class, but want to confirm.

And while I'm asking, is "Perform loading outside of this class" something that's already happening to `TunerModel`, or is that a TODO item for all the source options?

**Comment by DSheirer** [comment]
Hi @Sammy1Am.  

That TODO should have been cleaned up during the last redesign when I switched half of the program to use JavaFX.  

The P25 decoders (and soon DMR) probably won't work well with piped in audio in the current configuration.  They're tuned for use with the DQPSK demodulators which require complex samples.

The MPT1327 decoder and any of the AFSK1200 in-band audio decoders were the only ones that worked well with piped in audio.

With that in mind, I made the decision to drop support for the piped-in audio.

Before you get too far down the road, is there a specific protocol you're looking to leverage for piped in audio?

**Comment by Sammy1Am** [comment]
Well, it's a little complicated, but my motivation was basically three-fold:

- Since I don't (yet) have a specific station in mind to decode, it's a lot easier to just quickly scroll though frequencies in SDR Console and see which ones are something decode-able (I'm getting better a recognizing different types of signal, but it's not always immediately clear if they're clear enough to decode).  Piping audio from SDR Console makes that easier.
- I have an SDRPlay radio, which isn't yet supported by sdrtrunk, so audio would have allowed me to use my radio.
- DMR is actually pretty high on my list for protocols, but my top priority would probably be Motorola trunking systems (which most of the public safety radio around me uses).

I figured audio piping would be a good place to start so I could try out sdrtrunk with my radio and DMR signals before I got too ambitious and tried to add support for SDRPlay radios or Motorola protocols.  Looking at the DMR commits, I think trying to add support for Motorola might be a bit beyond me (plenty of Java/coding experience, but not sure even where to start with programmatically decoding radio signals).  I might see if I can figure out how to add support for the SDRPlay radios instead.  My understanding is they use a service that provides an API, so I may not need to get into USB nitty-gritty too much.

I'm excited to have found sdrtrunk though, it's nice to see some SDR software still getting updates, and open source to boot. 
 Unitrunker is the only program that seems to support the Motorola protocols, but it also doesn't support SDRPlay or audio (anymore).  DSDPlus supports audio (and by extension SDRPlay), but doesn't do Motorola and the free version hasn't been updated in ages.

**Comment by DSheirer** [comment]
Support for SDRPlay radios is planned for the 0.5.0 release, but I'll finish the DMR decoder before I start on that.

Adding support for the Motorola 3600 protocol would be much simpler than DMR.  DMR is by far the most complex protocol I've had to add so far, simply because there are so many flavors of DMR out in the wild.

**Comment by ah1102** [comment]
Danny, can you add AM / FM analog noise canceling to your software after the DMR decoder is complete?

**Comment by Sammy1Am** [comment]
Thanks for the reply, I got a bit busy and wasn't doing anything SDR for a while.  I was scanning around a bit recently and actually there's a lot of P25 in my area, and having failed to get OP25 setup, I'm back looking at sdrtrunk again.

With piped-in audio out (for good reason it sounds like), I'm left with getting SDRplay support so I've started in on that (#933).  Going smoothly so far, but I haven't gotten too far into the weeds so I'm sure I have plenty of learning ahead.
